### ISSUE TITLE
nnUNet training all

### ISSUE THREAD

**Issue by tchoi1991** [MIC-DKFZ/nnUNet#1364]
Hello Fabian,

First of all, thank you for your work and I'm truly enjoying using no new Net.

I have one small question, I trained with the following commands:

**nnUNet_train 3d_fullres nnUNetTrainerV2 901 all**

Basically, I trained all folds, and I got the 'all' directory in the 'result'.
I guess nnUNet validating all datasets right now. 

So is that mean when I run all training, they automatically train all folds at the same time and
they compute the metric(dice score) against all folds? like the mean of the dice score of all folds.

If I missed the instruction that you provided, please let me know.

Thank you!

**Comment by FabianIsensee** [comment]
Hey, 

nnU-net supports two modes:
- cross-validation: you need to run folds 0, 1, 2, 3 and 4. Each fold will have a proper validation set so that you can estimate the models performance on the training set. Inference is typically done using the 5 resulting models as an ensemble
- 'train on all data', abbreviated with the 'all' fold identifier. So there is no validation set! nnU-net takes ALL training cases and trains a single model on them. Useful for when you just want one model to do inference with

I hope this makes is clearer?
Best,
Fabian

**Comment by dscarmo** [comment]
> Hey,
> 
> nnU-net supports two modes:
> 
> * cross-validation: you need to run folds 0, 1, 2, 3 and 4. Each fold will have a proper validation set so that you can estimate the models performance on the training set. Inference is typically done using the 5 resulting models as an ensemble
> * 'train on all data', abbreviated with the 'all' fold identifier. So there is no validation set! nnU-net takes ALL training cases and trains a single model on them. Useful for when you just want one model to do inference with
> 
> I hope this makes is clearer? Best, Fabian

Hello Isensee, if there is no validation, why does when I train with the "all" fold there is a validation report on the progress reports? Is the validation being performed on the training samples? 

Thank you

**Comment by rongzhao-zhang** [comment]
When using 'all' fold, training set is the same as validation set.
you can check this [link](https://github.com/MIC-DKFZ/nnUNet/blob/nnunetv1/nnunet/training/network_training/nnUNetTrainerV2.py#L287)
```
if self.fold == "all":
        # if fold==all then we use all images for training and validation
        tr_keys = val_keys = list(self.dataset.keys())
```

**Comment by FabianIsensee** [comment]
Indeed, thanks @rongzhao-zhang

**Comment by dscarmo** [comment]
> > Hey,
> > nnU-net supports two modes:
> > ```
> > * cross-validation: you need to run folds 0, 1, 2, 3 and 4. Each fold will have a proper validation set so that you can estimate the models performance on the training set. Inference is typically done using the 5 resulting models as an ensemble
> > 
> > * 'train on all data', abbreviated with the 'all' fold identifier. So there is no validation set! nnU-net takes ALL training cases and trains a single model on them. Useful for when you just want one model to do inference with
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I hope this makes is clearer? Best, Fabian
> 
> Hi, I do not understand then how do you choose best checkpoint if `-f all`, i.e. there is no validation set? Is `checkpoint-best.pth` saved just the last checkpoint? This could overfit train for the default number of epochs which 1000.

I believe the overfit won't be that strong because of the very heavy data augmentation nnunet uses (check the supplementary material)

But in any case, I believe if you want to have a separate validation set holdout style you can create a custom split with a single fold and just train that split, thats what I am doing.

**Comment by dscarmo** [comment]
> > ,
> 
> You're right about creating custom split. But still, despite the augmentation, to me training and not having any validation set as an option doesn't make sense. For example, how is the `checkpoint-best.pth` determined in this case? Is it just the final checkpoint?

I did experiment with the 'all' option and surprisingly the test results for me are decent and comparable to using a custom validation split. The best validation loss for the checkpoint was pretty close to the last epoch but not necessarily the last one, probably due to augmentation. However, I would still not recommend doing that for fear of overfitting for sure.
